vms/evm/uptimetracker/uptime_tracker_test.go

```diff
@@ -0,0 +1,1045 @@
+	require.NoError(ut.Shutdown())
+
+	// Create new UptimeTracker from same DB with same validator set
+	newUT := setupUptimeTracker(t, validatorSet, db)
+
+	// Verify validator was loaded from DB
+	validator, exists := newUT.GetValidator(vID)
+	require.True(exists)
+	require.Equal(nodeID, validator.NodeID)
+	require.Equal(vID, validator.ValidationID)
+
+	// Verify uptime was loaded from DB
+	uptime, lastUpdated, err := newUT.GetUptime(nodeID)
+	require.NoError(err)
+	require.Equal(newUptime, uptime)
+	require.Equal(newLastUpdated.Unix(), lastUpdated.Unix())
+}
+
+func TestUptimeTracker_MultipleValidators(t *testing.T) {
+	require := require.New(t)
+
+	nodeID1 := ids.GenerateTestNodeID()
+	nodeID2 := ids.GenerateTestNodeID()
+	vID1 := ids.GenerateTestID()
+	vID2 := ids.GenerateTestID()
+
+	// Add multiple validators
+	validatorSet := map[ids.ID]*validators.GetCurrentValidatorOutput{
+		vID1: {
+			ValidationID:  vID1,
+			NodeID:        nodeID1,
+			Weight:        1000,
+			StartTime:     0,
+			IsActive:      true,
+			IsL1Validator: false,
+		},
+		vID2: {
+			ValidationID:  vID2,
+			NodeID:        nodeID2,
+			Weight:        2000,
+			StartTime:     0,
+			IsActive:      true,
+			IsL1Validator: true,
+		},
+	}
+
+	ut := setupUptimeTracker(t, validatorSet, nil)
+	require.NoError(ut.Sync(context.Background()))
+
+	// Verify both validators were added
+	validatorList := ut.GetValidators()
+	require.Len(validatorList, 2)
+
+	// Verify first validator
+	validator1, exists := ut.GetValidator(vID1)
+	require.True(exists)
+	require.Equal(nodeID1, validator1.NodeID)
+	require.Equal(uint64(1000), validator1.Weight)
+	require.False(validator1.IsL1Validator)
+
+	// Verify second validator
+	validator2, exists := ut.GetValidator(vID2)
+	require.True(exists)
+	require.Equal(nodeID2, validator2.NodeID)
+	require.Equal(uint64(2000), validator2.Weight)
+	require.True(validator2.IsL1Validator)
+
+	// Update one validator
+	validatorSet[vID1].Weight = 1500
+	validatorSet[vID1].IsActive = false
+	require.NoError(ut.Sync(context.Background()))
+
+	// Verify update
+	validator1, exists = ut.GetValidator(vID1)
+	require.True(exists)
+	require.Equal(uint64(1500), validator1.Weight)
+	require.False(validator1.IsActive)
+
+	// Remove one validator
+	delete(validatorSet, vID1)
+	require.NoError(ut.Sync(context.Background()))
+
+	// Verify removal
+	validatorList = ut.GetValidators()
+	require.Len(validatorList, 1)
+	require.Equal(vID2, validatorList[0].ValidationID)
+}
+
+// testValidatorState is a minimal implementation that only provides what we actually use
+type testValidatorState struct {
+	validators map[ids.ID]*validators.GetCurrentValidatorOutput
+}
+
+func (t *testValidatorState) GetCurrentValidatorSet(_ context.Context, _ ids.ID) (map[ids.ID]*validators.GetCurrentValidatorOutput, uint64, error) {
+	return t.validators, uint64(len(t.validators)), nil
+}
+
+// Stub implementations for the rest of the interface (unused in our tests)
```

Comment:
taking suggestions on how to get rid of this / `testValidatorState` if possible.